File: backend/app/schemas/scenario.py
"""Scenario API schemas."""

from pydantic import BaseModel, Field

from app.schemas.supplier import SupplierScoreSchema


class ScenarioCreateRequest(BaseModel):
    project_id: str
    name: str = "Custom Scenario"
    description: str | None = None
    shipping_multiplier: float = 1.0
    currency_rate: float = 1.0
    demand_multiplier: float = 1.0
    lead_time_adjustment_days: int = 0
    disabled_supplier_ids: list[str] = Field(default_factory=list)


class ScenarioSummarySchema(BaseModel):
    id: str
    project_id: str
    name: str
    description: str | None = None
    shipping_multiplier: float
    currency_rate: float
    demand_multiplier: float
    lead_time_adjustment_days: int
    disabled_supplier_ids: list[str] = Field(default_factory=list)


class ScenarioRankingDeltaSchema(BaseModel):
    supplier_id: str
    supplier_name: str
    baseline_rank: int
    scenario_rank: int
    rank_changed: bool
    baseline_score: float
    scenario_score: float
    landed_cost: float
    scores: SupplierScoreSchema


class ScenarioSimulationSchema(BaseModel):
    scenario_id: str
    previous_top_supplier_id: str
    new_top_supplier_id: str
    ranking_changed: bool
    rankings: list[ScenarioRankingDeltaSchema] = Field(default_factory=list)
    # Phase 6: AI-generated explanation of the ranking change
    explanation: str | None = None

File: frontend/src/types/scenario.ts
import type { SupplierScore } from './supplier';

export interface ScenarioCreateRequest {
  project_id: string;
  name: string;
  description?: string | null;
  shipping_multiplier: number;
  currency_rate: number;
  demand_multiplier: number;
  lead_time_adjustment_days: number;
  disabled_supplier_ids: string[];
}

export interface ScenarioSummary {
  id: string;
  project_id: string;
  name: string;
  description: string | null;
  shipping_multiplier: number;
  currency_rate: number;
  demand_multiplier: number;
  lead_time_adjustment_days: number;
  disabled_supplier_ids: string[];
}

export interface ScenarioRankingDelta {
  supplier_id: string;
  supplier_name: string;
  baseline_rank: number;
  scenario_rank: number;
  rank_changed: boolean;
  baseline_score: number;
  scenario_score: number;
  landed_cost: number;
  scores: SupplierScore;
}

export interface ScenarioSimulation {
  scenario_id: string;
  previous_top_supplier_id: string;
  new_top_supplier_id: string;
  ranking_changed: boolean;
  rankings: ScenarioRankingDelta[];
  explanation: string | null;
  /** Shipping multiplier that was applied — used by SensitivityChart */
  scenario_shipping_multiplier?: number;
}
